models/classes/counter/CounterService.php

```diff
@@ -0,0 +1,195 @@
+use oat\oatbox\service\ConfigurableService;
+use oat\oatbox\service\exception\InvalidServiceManagerException;
+use oat\oatbox\service\ServiceNotFoundException;
+
+class CounterService extends ConfigurableService
+{
+    public const SERVICE_ID = 'tao/CounterService';
+
+    public const OPTION_PERSISTENCE = 'persistence';
+    public const OPTION_COUNTER_KEY_PREFIX = 'counterKeyPrefix';
+    public const OPTION_EVENTS = 'events';
+
+    public const DEFAULT_PREFIX = 'tao:counter:';
+
+    protected const OPTION_CALLBACK = 'keyCallbackMethod';
+    protected const OPTION_SHORT_NAME = 'shortName';
+
+    /**
+     * @return KeyValuePersistence
+     * @throws CounterServiceException
+     */
+    protected function getPersistence(): KeyValuePersistence
+    {
+        $persistenceId = $this->getOption(self::OPTION_PERSISTENCE);
+        $persistence = KeyValuePersistence::getPersistence($persistenceId);
+
+        if (!$persistence instanceof KeyValuePersistence) {
+            $msg = "Persistence '${persistenceId}' must be an instance of '";
+            $msg .= KeyValuePersistence::class . "', ";
+            $msg .= get_class($persistence) . ' persistence given.';
+            throw new CounterServiceException($msg, CounterServiceException::CODE_INVALID_PERSISTENCE);
+        }
+
+        return $persistence;
+    }
+
+    /**
+     * @param string      $eventFqcn
+     * @param string      $shortName
+     * @param string|null $keyCallbackMethod
+     *
+     * @throws CounterServiceException
+     * @throws InvalidServiceManagerException
+     * @throws ServiceNotFoundException
+     * @throws common_Exception
+     */
+    public function attach(string $eventFqcn, string $shortName, ?string $keyCallbackMethod = null): void
+    {
+        if (!class_exists($eventFqcn)) {
+            $msg = "No event class found with FQCN '${$eventFqcn}' in available code base.";
+            throw new CounterServiceException($msg, CounterServiceException::CODE_UNKNOWN_EVENT_TYPE);
+        }
+
+        /** @var EventManager $eventManager */
+        $eventManager = $this->getServiceManager()->get(EventManager::SERVICE_ID);
+        $eventManager->attach($eventFqcn, [self::SERVICE_ID, 'increment']);
+        $this->getServiceManager()->register(EventManager::SERVICE_ID, $eventManager);
+
+        $eventOptions = $this->getOption(self::OPTION_EVENTS);
+        $eventOptions[$eventFqcn] = [
+            self::OPTION_CALLBACK => $keyCallbackMethod,
+            self::OPTION_SHORT_NAME => $shortName
+        ];
+        $this->setOption(self::OPTION_EVENTS, $eventOptions);
+
+        $this->logDebug("CounterService now listening to events with FQCN '${eventFqcn}'");
+    }
+
+    /**
+     * @param string $eventFqcn
+     *
+     * @throws InvalidServiceManagerException|CounterServiceException
+     * @throws \oat\oatbox\service\ServiceNotFoundException
+     * @throws \common_Exception
+     */
+    public function detach(string $eventFqcn): void
```

Comment:
Suggestion: I would avoid redundant annotations and also import FGN

```suggestion
    /**
     * @throws InvalidServiceManagerException|CounterServiceException
     * @throws ServiceNotFoundException
     * @throws common_Exception
     */
    public function detach(string $eventFqcn): void
```